src/map.cpp

```diff
@@ -529,62 +529,138 @@ bool Map::canThrowObjectTo(const Position& fromPos, const Position& toPos, bool
 	return isSightClear(fromPos, toPos, false);
 }
 
-bool Map::checkSightLine(const Position& fromPos, const Position& toPos) const
+bool Map::isTileClear(uint16_t x, uint16_t y, uint8_t z, bool blockFloor /*= false*/) const
 {
-	if (fromPos == toPos) {
+	const Tile* tile = getTile(x, y, z);
+	if (!tile) {
 		return true;
 	}
 
-	Position start(fromPos.z > toPos.z ? toPos : fromPos);
-	Position destination(fromPos.z > toPos.z ? fromPos : toPos);
-
-	const int8_t mx = start.x < destination.x ? 1 : start.x == destination.x ? 0 : -1;
-	const int8_t my = start.y < destination.y ? 1 : start.y == destination.y ? 0 : -1;
+	if (blockFloor && tile->getGround()) {
+		return false;
+	}
 
-	int32_t A = Position::getOffsetY(destination, start);
-	int32_t B = Position::getOffsetX(start, destination);
-	int32_t C = -(A * destination.x + B * destination.y);
+	return !tile->hasProperty(CONST_PROP_BLOCKPROJECTILE);
+}
 
-	while (start.x != destination.x || start.y != destination.y) {
-		int32_t move_hor = std::abs(A * (start.x + mx) + B * (start.y) + C);
-		int32_t move_ver = std::abs(A * (start.x) + B * (start.y + my) + C);
-		int32_t move_cross = std::abs(A * (start.x + mx) + B * (start.y + my) + C);
+bool Map::getSteepLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	float dx = x1 - x0;
+	float dy = y1 - y0;
+	float grad = (dx == 0) ? 1 : dy / dx;
+	float xy = y0 + grad;
 
-		if (start.y != destination.y && (start.x == destination.x || move_hor > move_ver || move_hor > move_cross)) {
-			start.y += my;
+	for (uint16_t y = x0 + 1; y < x1; ++y) {
+		if (!isTileClear(std::floor(xy), y, z)) {
+			return false;
 		}
+		xy += grad;
+	}
 
-		if (start.x != destination.x && (start.y == destination.y || move_ver > move_hor || move_ver > move_cross)) {
-			start.x += mx;
-		}
+	return true;
+}
+
+bool Map::getSlightLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	float dx = x1 - x0;
+	float dy = y1 - y0;
+	float grad = (dx == 0) ? 1 : dy / dx;
+	float xy = y0 + grad;
 
-		const Tile* tile = getTile(start.x, start.y, start.z);
-		if (tile && tile->hasProperty(CONST_PROP_BLOCKPROJECTILE)) {
+	for (uint16_t x = x0 + 1; x < x1; ++x) {
+		if (!isTileClear(x, std::floor(xy + 0.1), z)) {
```

Comment:
```suggestion
		if (!isTileClear(x, std::floor(yi + 0.1), z)) {
```

Again the magic number surfaces. Did you test with 0.5 instead of 0.1? Or using `std::round` instead of `std::floor`?

Reply:
with `std::round` (no "magic numbers") it behaves same as Bresenham's algo - the lines are completely different, vastly limiting our ability to throw diagonally near obstacles

with `std::floor` (no "magic numbers"), I don't experience the ["bug"](https://github.com/otland/forgottenserver/pull/3223#issuecomment-739238206) at all, but it might be compiler dependent

Reply:
I found the cause of the rounding error. It's the loss of precition at large positions (32k~). This is why I wasn't able to reproduce it earlier.

![obraz](https://user-images.githubusercontent.com/5205079/124353229-90f73780-dc05-11eb-81db-9c0d3d79d2f0.png)